src/primitives.jl

```diff
@@ -0,0 +1,425 @@
+#                                Heatmap, Image                                #
+################################################################################
+
+function draw_atomic(scene::Scene, screen::CairoScreen, primitive::Union{Heatmap, Image})
+    ctx = screen.context
+    image = primitive[3][]
+    x, y = primitive[1][], primitive[2][]
+    model = primitive[:model][]
+    imsize = (extrema_nan(x), extrema_nan(y))
+    xy_ = project_position(scene, Point2f0(first.(imsize)), model)
+    xymax_ = project_position(scene, Point2f0(last.(imsize)), model)
+    xy = min.(xy_, xymax_)
+    xymax = max.(xy_, xymax_)
+    w, h = xymax .- xy
+    interp = to_value(get(primitive, :interpolate, true))
+    # TODO: use Gaussian blurring
+    interp = interp ? Cairo.FILTER_BEST : Cairo.FILTER_NEAREST
+    s = to_cairo_image(image, primitive)
+    Cairo.rectangle(ctx, xy..., w, h)
+    Cairo.save(ctx)
+    Cairo.translate(ctx, xy[1], xy[2])
+    Cairo.scale(ctx, w / s.width, h / s.height)
+    Cairo.set_source_surface(ctx, s, 0, 0)
+    p = Cairo.get_source(ctx)
+    # Set filter doesn't work!?
+    Cairo.pattern_set_filter(p, interp)
+    Cairo.fill(ctx)
+    Cairo.restore(ctx)
+end
+
+################################################################################
+#                                     Mesh                                     #
+################################################################################
+
+function draw_atomic(scene::Scene, screen::CairoScreen, primitive::AbstractPlotting.Mesh)
+    @get_attribute(primitive, (color,))
+
+    colormap = get(primitive, :colormap, nothing) |> to_value |> to_colormap
+    colorrange = get(primitive, :colorrange, nothing) |> to_value
+
+    ctx = screen.context
+    model = primitive.model[]
+    mesh = primitive[1][]
+    vs = coordinates(mesh); fs = faces(mesh)
+    uv = hasproperty(mesh, :uv) ? mesh.uv : nothing
+    pattern = Cairo.CairoPatternMesh()
+
+    cols = per_face_colors(color, colormap, colorrange, vs, fs, uv)
+    for (f, (c1, c2, c3)) in zip(fs, cols)
+        t1, t2, t3 =  project_position.(scene, vs[f], (model,)) #triangle points
+        Cairo.mesh_pattern_begin_patch(pattern)
+
+        Cairo.mesh_pattern_move_to(pattern, t1...)
+        Cairo.mesh_pattern_line_to(pattern, t2...)
+        Cairo.mesh_pattern_line_to(pattern, t3...)
+
+        mesh_pattern_set_corner_color(pattern, 0, c1)
+        mesh_pattern_set_corner_color(pattern, 1, c2)
+        mesh_pattern_set_corner_color(pattern, 2, c3)
+
+        Cairo.mesh_pattern_end_patch(pattern)
+    end
+    Cairo.set_source(ctx, pattern)
+    Cairo.close_path(ctx)
+    Cairo.paint(ctx)
+    return nothing
+end
+
+################################################################################
+#                    Poly - the not so primitive, primitive                    #
+################################################################################
```

Comment:
I think either the file should not be called `primitives` but `plotobjects`, if there are other things than `AbstractPlotting` primitives in there, or you move this part to a separate file. Maybe the second version, there could be other things theoretically, who knows.

Reply:
I guess we should separate recipe overrides from real "primitives", so I'll just move poly to a separate file.